### PR TITLE
Security implementation

### PR DESCRIPTION
**Pull request by liubcheck**

### CODE REVIEW

File: src/main/java/liubomyr/stepanenko/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
+@RequiredArgsConstructor
+public class SecurityConfig {
+    private final UserDetailsService userDetailsService;
```
Comment (sokimaaa) [inline src/main/java/liubomyr/stepanenko/bookstore/config/SecurityConfig.java]: wrong way to inject beans in @Configuration. 
explore the abilities of @Bean.
```suggestion
public class SecurityConfig {
```

File: src/main/java/liubomyr/stepanenko/bookstore/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain getSecurityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/api/auth/*", "/swagger-ui/**")
```
Comment (sokimaaa) [inline src/main/java/liubomyr/stepanenko/bookstore/config/SecurityConfig.java]: permit error page as well.

File: src/main/java/liubomyr/stepanenko/bookstore/model/User.java [hunk 3]
```diff
@@ -28,6 +34,43 @@ public class User {
...
     private String lastName;
     private String shippingAddress;
+    @OneToMany
+    private Set<Role> roles;
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of();
```
Comment (sokimaaa) [inline src/main/java/liubomyr/stepanenko/bookstore/model/User.java]: any default authorities for user?

File: src/main/java/liubomyr/stepanenko/bookstore/repository/role/RoleRepository.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package liubomyr.stepanenko.bookstore.repository.role;
+
+import liubomyr.stepanenko.bookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (sokimaaa) [inline src/main/java/liubomyr/stepanenko/bookstore/repository/role/RoleRepository.java]: any sense
```suggestion
```

File: src/main/java/liubomyr/stepanenko/bookstore/security/CustomUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
```
Comment (sokimaaa) [inline src/main/java/liubomyr/stepanenko/bookstore/security/CustomUserDetailsService.java]: make exception unchecked. 
```suggestion
    public UserDetails loadUserByUsername(String email) {
```

File: src/main/java/liubomyr/stepanenko/bookstore/model/Role.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+@Entity
+@Table(name = "roles")
+@Data
+public class Role {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true, columnDefinition = "varchar")
+    private RoleName name;
```
Comment (Rommelua) [inline src/main/java/liubomyr/stepanenko/bookstore/model/Role.java]:
```suggestion
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, unique = true)
    private RoleName name;
```

File: src/main/java/liubomyr/stepanenko/bookstore/model/User.java [hunk 7]
```diff
@@ -28,6 +35,43 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @OneToMany
```
Comment (Rommelua) [inline src/main/java/liubomyr/stepanenko/bookstore/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/liubomyr/stepanenko/bookstore/model/User.java [hunk 8]
```diff
@@ -28,6 +35,43 @@ public class User {
...
+    private Set<Role> roles;
     @Column(nullable = false)
     private boolean isDeleted = false;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of(new SimpleGrantedAuthority("ROLE_USER"));
```
Comment (Rommelua) [inline src/main/java/liubomyr/stepanenko/bookstore/model/User.java]:
```suggestion
        return roles.stream()
                .map(r -> new SimpleGrantedAuthority("ROLE_" + r.getRoleName().name()))
                .collect(Collectors.toList());
```

File: src/main/java/liubomyr/stepanenko/bookstore/model/enums/RoleName.java [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package liubomyr.stepanenko.bookstore.model.enums;
```
Comment (Rommelua) [inline src/main/java/liubomyr/stepanenko/bookstore/model/enums/RoleName.java]:
```suggestion
package liubomyr.stepanenko.bookstore.model.enum;
```

**Pull request merged**